frontend/packages/sdk/src/contracts/PinkBlueprint.ts

```diff
@@ -0,0 +1,309 @@
+import type { ApiPromise } from '@polkadot/api';
+import type { KeyringPair } from '@polkadot/keyring/types';
+
+import type { OnChainRegistry } from '../OnChainRegistry';
+import type { InkResponse, InkQueryError, AbiLike } from '../types';
+import type { CertificateData } from '../certificate';
+
+import { SubmittableResult } from '@polkadot/api';
+import { ApiBase } from '@polkadot/api/base';
+import { BN_ZERO, isUndefined, hexAddPrefix, u8aToHex, hexToU8a } from '@polkadot/util';
+import { createBluePrintTx, withMeta } from '@polkadot/api-contract/base/util';
+import { sr25519Agree, sr25519KeypairFromSeed, sr25519Sign } from "@polkadot/wasm-crypto";
+import crypto from 'crypto';
+import { from } from 'rxjs';
+
+import { Abi } from '@polkadot/api-contract/Abi';
+import { toPromiseMethod } from '@polkadot/api';
+import { Option } from '@polkadot/types';
+
+import { pruntime_rpc as pruntimeRpc } from "../proto";
+import { decrypt, encrypt } from "../lib/aes-256-gcm";
+import { randomHex } from "../lib/hex";
+import { PinkContractPromise } from './PinkContract';
+
+
+interface ContractInkQuery<ApiType extends ApiTypes> extends MessageMeta {
+  (origin: KeyringPair, ...params: unknown[]): ContractCallResult<ApiType, ContractCallOutcome>;
+}
+
+interface MapMessageInkQuery<ApiType extends ApiTypes> {
+  [message: string]: ContractInkQuery<ApiType>;
+}
+
+interface PinkContractInstantiateResult extends ContractInstantiateResult {
+  salt: string;
+}
+
+function hex(b: string | Uint8Array) {
+  if (typeof b != "string") {
+    b = Buffer.from(b).toString('hex');
+  }
+  if (!b.startsWith('0x')) {
+    return '0x' + b;
+  } else {
+    return b;
+  }
+}
+
+function createEncryptedData(pk: Uint8Array, data: string, agreementKey: Uint8Array) {
+  const iv = hexAddPrefix(randomHex(12));
+  return {
+    iv,
+    pubkey: u8aToHex(pk),
+    data: hexAddPrefix(encrypt(data, agreementKey, hexToU8a(iv))),
+  };
+};
+
+async function pinkQuery(
+  api: ApiPromise,
+  pruntimeApi: pruntimeRpc.PhactoryAPI,
+  pk: Uint8Array,
+  queryAgreementKey: Uint8Array,
+  encodedQuery: string,
+  { certificate, pubkey, secret }: CertificateData
+) {
+  // Encrypt the ContractQuery.
+  const encryptedData = createEncryptedData(pk, encodedQuery, queryAgreementKey);
+  const encodedEncryptedData = api
+    .createType("EncryptedData", encryptedData)
+    .toU8a();
+
+  // Sign the encrypted data.
+  const signature: pruntimeRpc.ISignature = {
+    signedBy: certificate,
+    signatureType: pruntimeRpc.SignatureType.Sr25519,
+    signature: sr25519Sign(pubkey, secret, encodedEncryptedData),
+  };
+
+  // Send request.
+  const requestData = {
+    encodedEncryptedData,
+    signature,
+  };
+  return pruntimeApi.contractQuery(requestData).then((res) => {
+    const { encodedEncryptedData } = res;
+    const { data: encryptedData, iv } = api
+      .createType("EncryptedData", encodedEncryptedData)
+      .toJSON() as {
```

Comment:
Prefer using the polkadot object directly (`toHex()` or `u8aToHex()` when necessary) than `toJSON()` and `toHuman()`. The latter methods are not lossless.